golang/cosmos/vm/action.go

```diff
@@ -0,0 +1,104 @@
+package vm
+
+import (
+	"reflect"
+	"strconv"
+	"time"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+var (
+	zeroTime         = time.Time{}
+	actionHeaderType = reflect.TypeOf(ActionHeader{})
+)
+
+// Jsonable is a value, j, that can be passed through json.Marshal(j).
+type Jsonable interface{}
+
+// ActionHeader must be embedded in all actions.  It is populated by PopulateAction.
+type ActionHeader struct {
+	// Type defaults to the `default:"..."` tag of the embedder's ActionHeader field.
+	Type string `json:"type"`
+	// BlockHeight defaults to sdk.Context.BlockHeight().
+	BlockHeight int64 `json:"blockHeight"`
+	// BlockTime defaults to sdk.Context.BlockTime().Unix().
+	BlockTime int64 `json:"blockTime"`
+}
+
+// ActionPusher enqueues data for later consumption by the controller.
+type ActionPusher func(ctx sdk.Context, action Jsonable) error
+
+func setActionHeaderDefaults(ctx sdk.Context, defaultType string, ah *ActionHeader) {
+	// Default the Type field to the default tag.
+	if len(defaultType) != 0 && len(ah.Type) == 0 {
+		ah.Type = defaultType
+	}
```

Comment:
Is there a case where `defaultType` would not be set?

Should we assert that `Type` is never empty?